### PR TITLE
[DEV-1022] Editable profile

### PR DESCRIPTION
**Pull request by tommaso1**
#### List of Changes

Editable profile page

#### Motivation and Context

DEV-1022

#### How Has This Been Tested?

go to `/profile/personal-data`, then click "edit"

#### Screenshots (if appropriate):
proof

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: apps/nextjs-website/src/app/profile/personal-data/page.tsx [hunk 1]
```diff
@@ -14,37 +15,48 @@ const PersonalData = () => {
...
+        },
+        {
+          title: t('personalData.fields.sector'),
+          value: user?.attributes['custom:company_type'],
+          editable: true,
+        },
+      ]
+    )
+    console.log('user?.attributes', user?.attributes);
+  }, [user?.attributes])
```
Comment (marcobottaro) [inline apps/nextjs-website/src/app/profile/personal-data/page.tsx:48]: This file has linting issues

File: apps/nextjs-website/src/app/profile/personal-data/page.tsx [hunk 2]
```diff
@@ -14,37 +15,48 @@ const PersonalData = () => {
...
+          value: user?.attributes.family_name,
+          editable: true,
+        },
+        {
+          title: t('personalData.fields.role'),
+          value: user?.attributes['custom:job_role'],
+          editable: true,
+        },
+        {
+          title: t('personalData.fields.sector'),
```
Comment (marcobottaro) [inline apps/nextjs-website/src/app/profile/personal-data/page.tsx:41]: This field should be e Select like in the signu up form
image
Figma:
image

**Review by marcobottaro — CHANGES_REQUESTED** [review]

**Review by marcobottaro — APPROVED** [review]

File: apps/nextjs-website/src/app/profile/personal-data/page.tsx [hunk 3]
```diff
@@ -57,6 +75,29 @@ const PersonalData = () => {
...
         items={dataSectionItems}
+        onValue={(items: InfoCardItemProps[]) => {
+          const oldItems = [...dataSectionItems];
+
+          if (!user) return;
+
+          setUserAttributes(
+            {
+              ...user.attributes,
+              given_name: items[0].value || '',
```
Comment (marcobottaro) [inline apps/nextjs-website/src/app/profile/personal-data/page.tsx:86]: Maybe you could replace this and the following `|| ''` with something like this:
```suggestion
              ...(items[0].value && { given_name: items[0].value }),
```

**Review by MarcoPonchia — CHANGES_REQUESTED** [review]
User should access to edit form also from button "Inserisci" visible on [personal-data](http://localhost:3000/profile/personal-data) when there isn't any information relative to some field.
[Screencast from 14-12-2023 10:50:24.webm](https://github.com/pagopa/developer-portal/assets/24475319/d265121a-5f07-4fa8-8ccf-84d361475c40)

**Review by datalek — CHANGES_REQUESTED** [review]
There are some conflicts to resolve :/

**Review by MarcoPonchia — CHANGES_REQUESTED** [review]
I cannot save an empty field, as you can see in this clip when i save an empty field the previus value show up after save. I also think is bettere to align "name" and "surname" fields to signup form and make  theme required, if their empty form should show an error on relative input and save should be disabled
[Screencast from 17-01-2024 12:56:23.webm](https://github.com/pagopa/developer-portal/assets/24475319/4b7c7117-9224-4919-a14a-f06fcdba1470)

File: apps/nextjs-website/src/app/profile/personal-data/page.tsx [hunk 4]
```diff
@@ -37,36 +81,12 @@ const PersonalData = () => {
     setShowModal(true);
```
Comment (MarcoPonchia) [inline apps/nextjs-website/src/app/profile/personal-data/page.tsx:81]:
```suggestion
```

File: apps/nextjs-website/src/app/profile/personal-data/page.tsx [hunk 5]
```diff
@@ -24,8 +28,48 @@ const PersonalData = () => {
...
+        values: companyRoles,
+        required: false,
+      },
+    ]);
+  }, [user?.attributes]);
+
   const router = useRouter();
-  const { user } = useUser();
   const [editItem, setEditItem] = useState<InfoCardItemProps | null>(null);
   const [showModal, setShowModal] = useState(false);
```
Comment (MarcoPonchia) [inline apps/nextjs-website/src/app/profile/personal-data/page.tsx:74]:
```suggestion
```

File: apps/nextjs-website/src/app/profile/personal-data/page.tsx [hunk 6]
```diff
@@ -24,8 +28,48 @@ const PersonalData = () => {
...
+        value: user?.attributes['custom:company_type'],
+        editable: true,
+        type: 'select',
+        values: companyRoles,
+        required: false,
+      },
+    ]);
+  }, [user?.attributes]);
+
   const router = useRouter();
```
Comment (MarcoPonchia) [inline apps/nextjs-website/src/app/profile/personal-data/page.tsx:72]:
```suggestion
```

File: apps/nextjs-website/src/app/profile/personal-data/page.tsx [hunk 7]
```diff
@@ -108,30 +128,42 @@ const PersonalData = () => {
...
+
+            setUserAttributes(
+              {
+                ...user.attributes,
+                ...(items[0].value && { given_name: items[0].value }),
+                ...(items[1].value && { family_name: items[1].value }),
+                ...(items[2].value && { 'custom:job_role': items[2].value }),
+                ...(items[3].value && {
+                  'custom:company_type': items[3].value,
+                }),
```
Comment (MarcoPonchia) [inline apps/nextjs-website/src/app/profile/personal-data/page.tsx:152]:
```suggestion
                ...{
                  given_name: items[0].value || user.attributes.given_name,
                  family_name: items[1].value || user.attributes.family_name,
                  'custom:job_role': items[2].value || '',
                  'custom:company_type':
                    items[3].value || user.attributes['custom:company_type'],
                },
```

File: apps/nextjs-website/src/components/molecules/InfoCard/InfoCardProfile.tsx [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+            {...item}
+            valueFallback={addValueComponent}
+            editing={editing}
+            onValue={(value) => {
+              const newItems = [...dataSectionItems];
+              // eslint-disable-next-line functional/immutable-data
+              newItems[index] = {
+                ...newItems[index],
+                // workaround for cognito not accepting empty strings
+                value: value == '' ? ' ' : value,
```
Comment (MarcoPonchia) [inline apps/nextjs-website/src/components/molecules/InfoCard/InfoCardProfile.tsx:104]:
```suggestion
                value,
```

**Review by MarcoPonchia — CHANGES_REQUESTED** [review]

**Review by MarcoPonchia — APPROVED** [review]

**Comment by kin0992** [comment]
@tommaso1 please add the changeset in this PR

File: apps/nextjs-website/src/components/atoms/InfoCardItem/InfoCardItemProfile.tsx [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+        <ButtonNaked
+          onClick={infoCardItem.onInsertPressed}
+          color='primary'
+          sx={{ paddingLeft: 0, paddingRight: 0 }}
+        >
+          {t('insert')}
+        </ButtonNaked>
+      )}
+    </Stack>
+  );
```
Comment (MarcoPonchia) [inline apps/nextjs-website/src/components/atoms/InfoCardItem/InfoCardItemProfile.tsx:153]:
```suggestion
  if (!infoCardItem.editing) {
    return (
      <Stack
        my={{ xs: 1, md: 3 }}
        flexDirection={{ xs: 'column', md: 'row' }}
        alignItems={{ xs: 'flex-start', md: 'center' }}
        gap={2}
      >
        <Typography
          variant='body2'
          fontSize={16}
          minWidth={{ xs: 'auto', md: '170px' }}
        >
          {infoCardItem.title}
        </Typography>
        {infoCardItem.value && infoCardItem.value != '' ? (
          <Typography
            minHeight={'24px'}
            fontSize={16}
            flexGrow={1}
            fontWeight={700}
          >
            {infoCardItem.type === 'select'
              ? infoCardItem.values.find(
                  ({ value }) => value === infoCardItem.value
                )?.title
              : infoCardItem.value}
          </Typography>
        ) : (
          <ButtonNaked
            onClick={infoCardItem.onInsertPressed}
            color='primary'
            sx={{ paddingLeft: 0, paddingRight: 0 }}
          >
            {t('insert')}
          </ButtonNaked>
        )}
      </Stack>
    );
  }

  return (
    <Stack spacing={2} mb={2} sx={{ marginTop: '2rem' }}>
      <FormControl variant='outlined'>
        {infoCardItem.type === 'text' ? (
          infoCardItem.required ? (
            <RequiredTextField
              label={infoCardItem.title}
              value={infoCardItem.value ?? ''}
              onChange={({ target: { value } }) => {
                infoCardItem.onValue && infoCardItem.onValue(value);
              }}
              helperText={translations.shared.requiredFieldError}
            />
          ) : (
            <TextField
              variant='outlined'
              id={infoCardItem.title}
              type={'text'}
              onChange={({ target: { value } }) => {
                infoCardItem.onValue && infoCardItem.onValue(value);
              }}
              value={infoCardItem.value}
              label={infoCardItem.title}
              size='small'
            />
          )
        ) : (
          <>
            <InputLabel
              id={'company-field'}
              sx={{ backgroundColor: 'white', paddingX: '8px' }}
            >
              {infoCardItem.title}
              {infoCardItem.required ? '*' : ''}
            </InputLabel>
            <Select
              labelId='company-field'
              id='company-field-select'
              value={
                infoCardItem.values.find(
                  ({ value }) => value === infoCardItem.value
                )?.value || ''
              }
              onChange={({ target: { value } }) =>
                infoCardItem.onValue &&
                infoCardItem.onValue(
                  infoCardItem.values.find(({ value: v }) => v === value)
                    ?.value || ''
                )
              }
              sx={{ padding: '8.5px 14px' }}
              inputProps={{
                sx: {
                  padding: 0,
                },
              }}
            >
              {infoCardItem.values.map(({ title, value }) => (
                <MenuItem key={value} value={value}>
                  {title}
                </MenuItem>
              ))}
            </Select>
          </>
        )}
      </FormControl>
    </Stack>
  );
```

File: apps/nextjs-website/src/components/atoms/InfoCardItem/InfoCardItemProfile.tsx [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+import { translations } from '@/_contents/translations';
+import RequiredTextField from '@/components/molecules/RequiredTextField/RequiredTextField';
+import {
+  FormControl,
+  InputLabel,
+  MenuItem,
+  OutlinedInput,
+  Select,
+  Stack,
+  Typography,
```
Comment (MarcoPonchia) [inline apps/nextjs-website/src/components/atoms/InfoCardItem/InfoCardItemProfile.tsx:11]:
```suggestion
  FormControl,
  InputLabel,
  MenuItem,
  Select,
  Stack,
  TextField,
  Typography,
```

**Review by MarcoPonchia — CHANGES_REQUESTED** [review]

File: apps/nextjs-website/src/components/atoms/InfoCardItem/InfoCardItemProfile.tsx [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+              {infoCardItem.values.map(({ title, value }) => (
+                <MenuItem key={value} value={value}>
+                  {title}
+                </MenuItem>
+              ))}
+            </Select>
+          )}
+        </FormControl>
+      </Stack>
+    );
```
Comment (MarcoPonchia) [inline apps/nextjs-website/src/components/atoms/InfoCardItem/InfoCardItemProfile.tsx:114]:
```suggestion
  if (infoCardItem.editing) {
    return (
      <Stack spacing={2} mb={2} sx={{ marginTop: '2rem' }}>
        <FormControl variant='outlined'>
          {infoCardItem.type === 'text' ? (
            infoCardItem.required ? (
              <RequiredTextField
                label={infoCardItem.title}
                value={infoCardItem.value ?? ''}
                onChange={({ target: { value } }) => {
                  infoCardItem.onValue && infoCardItem.onValue(value);
                }}
                helperText={translations.shared.requiredFieldError}
              />
            ) : (
              <TextField
                variant='outlined'
                id={infoCardItem.title}
                type={'text'}
                onChange={({ target: { value } }) => {
                  infoCardItem.onValue && infoCardItem.onValue(value);
                }}
                value={infoCardItem.value}
                label={infoCardItem.title}
                size='small'
              />
            )
          ) : (
            <>
              <InputLabel
                id={'company-field'}
                sx={{
                  backgroundColor: 'white',
                  paddingX: '8px',
                  marginLeft: '-6px',
                }}
              >
                {infoCardItem.title}
                {infoCardItem.required ? '*' : ''}
              </InputLabel>
              <Select
                labelId='company-field'
                id='company-field-select'
                value={
                  infoCardItem.values.find(
                    ({ value }) => value === infoCardItem.value
                  )?.value || ''
                }
                onChange={({ target: { value } }) =>
                  infoCardItem.onValue &&
                  infoCardItem.onValue(
                    infoCardItem.values.find(({ value: v }) => v === value)
                      ?.value || ''
                  )
                }
                sx={{ padding: '8.5px 14px' }}
                inputProps={{
                  sx: {
                    padding: 0,
                  },
                }}
              >
                {infoCardItem.values.map(({ title, value }) => (
                  <MenuItem key={value} value={value}>
                    {title}
                  </MenuItem>
                ))}
              </Select>
            </>
          )}
        </FormControl>
      </Stack>
    );
  }
```

**Review by MarcoPonchia — CHANGES_REQUESTED** [review]

**Comment by marcobottaro** [comment]
When the sector value is missing the user doesn't see the "Insert" button:
image

**Comment by marcobottaro** [comment]
When the user edits the profile settings and the sector is empty, the placeholder is not aligned:
image

**Review by marcobottaro — CHANGES_REQUESTED** [review]
Look at the comments

**Comment by MarcoPonchia** [comment]
> When the sector value is missing the user doesn't see the "Insert" button: image

This case is not repeatable. I think is due to some old dirty data in your account not related to enum.

**Review by MarcoPonchia — APPROVED** [review]

**Review by marcobottaro — APPROVED** [review]

**Comment by kin0992** [comment]
in this PR a changeset would have been useful @MarcoPonchia @marcobottaro @tommaso1

**Pull request merged**